[WIP] implementation of `conan exec`

niosHD opened the pull request:
This is a WIP implementation of the `conan exec` feature which I proposed in #2461. It currently still lacks unit tests and has only been manually tested on Linux so far. However, at least for me, this prototype already works quite nicely.

Usage looks something like `conan exec [-pr PROFILE] [-ref REFERENCE] [...] COMMAND [ARGUMENTS ...]`. Note that also multiple references can be specified if desired. For example, executing `cmake --version` for cmake 3.5.2 looks then like `conan exec -ref cmake_installer/3.5.2@conan/stable cmake --version`.

Additionally, also all common `conan install` flags are supported which means that missing packages get installed automatically or can be built. Furthermore, customization of settings and options is possible.

Please give it a try and tell me what you think.

memsharded commented on conans/client/command.py:
```diff
@@ -1089,6 +1112,8 @@ def _commands(self):
             if not method_name.startswith('_'):
                 if "export_pkg" == method_name:
                     method_name = "export-pkg"
+                if "execute" == method_name:
```
I would say, to keep the command and method named exactly the same. The export-pkg is about consistency of underscore and dash, but otherwise, better be the same.

niosHD replied:
Sure, but we cannot use `exec` as command name then given that it is a reserved keyword in python.

memsharded commented on conans/client/conan_api.py:
```diff
@@ -681,6 +683,78 @@ def export_alias(self, reference, target_reference):
+        # Write all requested references to the file. Note that we have to
+        # close the file such that Windows can work with them.
+        # (see https://bugs.python.org/issue14243)
+        temp = tempfile.NamedTemporaryFile(suffix=".txt", delete=False)
```
I'd suggest to just create the full file contents as a string in memory, then save it with the ``conans.files.save()`` method.

niosHD replied:
I can do that but it would require changes to `conans.files.save()` given that the filename is generated temporary. Splitting the name generation into a call to `tempfile.mktemp` and `conans.files.save()` yields a potential race condition otherwise.

SSE4 replied:
the file is only needed for the self._manager.install, right?
may be it's better to provide an overload which accepts reference(s) instead of file name with reference(s)?

niosHD replied:
I absolutely agree, I would also prefer to implement this `exec` functionality completely without writing any temporary file. However, I didn't wanted to do a potential large refactoring of the manager.install code (which I am not really familiar with) only for a prototype. Furthermore, given that it is not even sure yet that the `exec` feature is something that will be added to conan, increasing the size of the change set would certainly not improve my chances to get accepted. :smile:

memsharded commented on conans/client/generators/virtualenv.py:
```diff
@@ -17,36 +17,52 @@ def __init__(self, conanfile):
+                    ret.append({'name': name, 'value': value, 'scriptvalue': '"' + value + '"'})
+        return ret
+
+    def _get_setenv_variables_commands(self):
```
Could you please clarify the changes of this method? It is not clear at all to the scope of this PR.

niosHD commented:
Just for completeness, I will obviously add unit tests in the following commits. I just wanted to get a first version out as soon as possible in order to get feedback early on and to keep the discussion from #2461 going.

For example, I am currently thinking about adding the possibility to specify a directory with an already existing conanfile.txt or .py instead of the references in order to provide an alternative solution for #2463. The idea would basically be to extend the virtual environment generator with the `CONAN_PACKAGE_ROOTS` variable using the code [you demonstrated](https://github.com/memsharded/cmake_find_generator).

The requested `$ PREFIX_PATH=$(conan install)` could then be implemented in the form of `$ PREFIX_PATH=$(conan exec --dir . echo $CONAN_PACKAGE_ROOTS)`. Or even better, given a small wrapper script for cmake which adds the environment variable to the command line, calling cmake could be done as `$ conan exec --dir . cmake_wapper <arguments>`.

SSE4 commented on conans/client/conan_api.py:
```diff
@@ -681,6 +683,78 @@ def export_alias(self, reference, target_reference):
+        for variable in env_vars:
+            name = variable['name']
+            value = variable['value']
+            if 'separator' in variable.keys() and name in new_env:
```
what's the magic separator environment variable?

niosHD replied:
It is simply the separator character which, if it is defined, has to be used to join the current value from the environment with the values from the virtual environment generator. In most cases it is simply `os.pathsep`. However, there are [some variables like CPPFLAGS](https://github.com/niosHD/conan/blob/develop/conans/client/generators/virtualenv.py#L8) that are joined with spaces by the virtual environment generator.

SSE4 commented on conans/client/generators/virtualenv.py:
```diff
@@ -17,36 +17,52 @@ def __init__(self, conanfile):
+        return ret
+
+    def _get_setenv_variables_commands(self):
+        command_set = "SET" if platform.system() == "Windows" else "export"
```
this sounds like a weak condition - always suppose that Windows implies CMD.EXE shell (it might be PowerShell, MSYS shell or Cygwin, for instance).

niosHD replied:
I agree, but this is the same as the [current implementation](https://github.com/niosHD/conan/blob/d508c44031542c95051d6036047ce156b935e8bc/conans/client/generators/virtualenv.py#L21) of the virtual environment generator. Commit 4376828d52896675de1091149cfc01d35aaaf503, which contains this change, is only a refactoring of the current code and should not alter the existing behavior in any way. I do not think that changing the generator output is in the scope of this PR.

niosHD commented:
I finally managed to continue with this PR such that we can continue the discussion. Given that I did not receive any more feedback from the last review round I did not change any naming yet. However, this PR still is a WIP/RFC version and all names and command line parameters are subject to change and merely a suggestion.

What has been changed since the last version:
* I rebased the PR on top of #2621 so the first two commits are not part of this PR.
* I added unit tests which show that my implementation actually works. The only reason CI is failing currently is because I still have problems with getting the stdout redirection right on all platforms when nose captures stdout with its Tee class. Running `nosetests` with `-s` works on Windows and Linux for me. Getting this right may still require a few additional iterations...
* I added support for suppressing output from conan with the `-q` option to enable uses like requested in #2463 (e.g., `$ <LOCAL_VARIABLE>=$(conan exec -ref <reference> printenv <variable>`). Note however that the requested directory paths are not exported into the environment at the moment and are subject for another PR.
* I added `--path` (instead of the mentioned `--dir`) such that `conan exec` can be used like `conan install` with file/directory path.
* I appended the `virtualrunenv` at the end of the `virtualenv` to enable the use of `conan exec` for situations like discussed in #2660.

Please give it a try and provide feedback on the code as well as on the overall idea.

lasote commented:
I still think Conan shouldn't support this command. It is difficult to reject this but it is important to keep the tool focused, the support of the million things Conan does is a bit insane already. Also AFAIK, this hasn't been demanded so far. Anyway thanks for your constant help to the project @niosHD
